### PR TITLE
Allow inline sessions without requiring the presence of an XR device

### PR DESCRIPTION
**Pull request by Manishearth**
Fixes https://github.com/immersive-web/webxr/issues/658, fixes  https://github.com/immersive-web/webxr/issues/635

r? @NellWaliczek

### CODE REVIEW

**Comment by ddorwin** [comment]
Please see my new https://github.com/immersive-web/webxr/issues/635#issuecomment-502292546.

More generally, I wonder if the existing "XR Device" concept, which doesn't apply to inline sessions, isn't prone to such issues. @toji

**Comment by Manishearth** [comment]
Personally I feel like for inline sessions we should support a pseudo-XR device or something.

This also gives us better control over defining views; for example if your browser is running in XR, the inline session may actually want to expose two views (and then composite the canvas correctly), but the concept of a view comes from the XR device.

File: index.bs [hunk 1]
```diff
@@ -316,9 +316,9 @@ When the <dfn method for="XR">requestSession(|mode|)</dfn> method is invoked, th
     1. [=ensures an XR device is selected|Ensure an XR device is selected=].
     1. Follow the steps for the first matching conditions:
         <dl class="switch">
-          <dt> If the [=XR/XR device=] is <code>null</code>
+          <dt> If the [=XR/XR device=] is <code>null</code> and |immersive| is <code>true</code>
           <dd> [=Reject=] |promise| with <code>null</code>.
-          <dt> If the [=XR/XR device=]'s [=list of supported modes=] does not [=list/contain=] |mode|
+          <dt> If |immersive| is not <code>true</code> and the [=XR/XR device=]'s [=list of supported modes=] does not [=list/contain=] |mode|
```
[the remark this answers is not in the document]
Comment (Manishearth) [reply index.bs]: I think this should be "If immersive is true and ...", the not somehow snuck in there.

We shouldn't reject immersive=false sessions

**Comment by Manishearth** [comment]
Addressed

File: index.bs (hunk 1 above)
[the remark this answers is not in the document]
Comment (toji) [reply index.bs]: Oof, sorry! You're right.

**Comment by Manishearth** [comment]
Rebased. I need to get better at not breaking my own PRs :stuck_out_tongue:

File: index.bs [hunk 2]
```diff
@@ -281,7 +281,7 @@ When the <dfn method for="XR">supportsSession(|mode|)</dfn> method is invoked, i
   1. Run the following steps [=in parallel=]:
     1. [=ensures an XR device is selected|Ensure an XR device is selected=].
     1. If the [=XR/XR device=] is null, [=reject=] |promise| with a "{{NotSupportedError}}" {{DOMException}} and abort these steps.
-    1. If the [=XR/XR device=]'s [=list of supported modes=] does not [=list/contain=] |mode|, [=reject=] |promise| with a "{{NotSupportedError}}" {{DOMException}} and abort these steps.
+    1. If the [=XR/XR device=]'s [=list of supported immersive modes=] does not [=list/contain=] |mode|, [=reject=] |promise| with a "{{NotSupportedError}}" {{DOMException}} and abort these steps.
```
Comment (NellWaliczek) [inline index.bs]: Maybe I'm missing something here... but won't this cause `supportsSession('inline')` to be rejected?

Comment (Manishearth) [reply index.bs]: Fixed

**Comment by toji** [comment]
> To be honest, I'm not sure landing this PR without the broader correction is super helpful. Am I missing a reason to try to land this one as-is...?

I'm assuming that we will not have an opportunity to fully address adding an inline XR device prior to the close of the June milestone, and this change would be nice to have in place to make sure our VR complete milestone demonstrates the desired behavior and gives us more leeway to update the implementation details after the fact.

**Comment by Manishearth** [comment]
I think we should land this first as a stopgap, but I can start playing with the idea of a default inline device afterwards.

**Comment by toji** [comment]
Nell asked me to do a sanity check and figure out how many places in the spec we're currently assuming an XR device will always be present when using an inline session. I have to admit, I was a bit surprised at the list:

 - 1.2 Application Flow:
   - Assumes the frame loop displays images on the XR Device
 - 4.1 XRSession:
   - Releasing exclusive access to the XR device.
   - Releasing graphics resources acquired by the XR device (N/A?)
   - Putting the XR device in a state that's receptive of new sessions.
   - Monitor input sources associated with the XR Device
   - Visibility state "visible" runs at refresh rate of XR device
   - Viewer reference space has list of views provided by the XR device
 - 4.3 Animation Frames:
   - Animation frame runs when updates to viewer state received from XR device
 - 6.1 XRSpace:
   - Native origin tracked by XR device's tracking system
 - 6.2 XRReferenceSpace:
   - Reference space supported algorithm makes several references to XR device
 - 7.1 XRView:
   - View corresnponds to display on XR device
   - Projection matrix provided by the underlying device
 - 9.2 XRViewerPose:
   - State of the viewer as tracked by the XR device
   - Every view must be rendered to correctly display on the XR device
 - 10.1 XRInputSource:
   - Input sources not associated with the XR device should not be returned
   - Input source may be removed from XR device before select is complete
 - 10.2 Transient input:
   - Some XR device's support transient input sources
 - 11.1 XRWebGLLayer:
   - Enabled presentation onto the XR device
   - Native framebuffer resolution matches XR device
 - 11.2 WebGL context compatibility:
   - Must be compatible with XR device to display imagery
   - xrCompatible flag algorithm makes compatible with XR device
   - makeXRCompatible flag algorithm makes compatible with XR device

Mind you that's ONLY places where the text in question may be applicable to inline sessions. There's a bunch of other places where the XR device is referenced but it's only applicable to immersive sessions, or the logic presented naturally allows for avoiding interaction with the XR device, and I haven't included them here.

And after dredging all of that up, I've got to say that I'm now on Nell's side of the fence on this one. If we allow for the possibility of no XR device then there's a bunch of logic that, given a strict reading, doesn't make sense. As such the cleaner solution is likely to just full spec out an inline-only XR device that's made available in the absense of actual XR hardware.

**Comment by Manishearth** [comment]
Sounds good, will do.

**Comment by Manishearth** [comment]
https://github.com/immersive-web/webxr/pull/737

**Pull request closed without merging**